contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+API version to match exactly the Kubelet's Device Plugin API version.
+Therefore if the Device Plugin API version change then you will have to change
+the Device Plugin too.
+
+
+*Future:*
+When the Device Plugin API becomes a stable feature, versionning should be
+backward compatible and even if Kubelet has a different Device Plugin API,
+it should not require a Device Plugin upgrade.
+
+Refer to the versioning section for versioning scheme compatibility.
+
+## Upgrading Device Plugins
+
+Because we cannot enforce what the different Device Plugins will do, we cannot
+say for certain that upgrading a device plugin will not crash any containers
+on the node.
+
+It is therefore up to the Device Plugin vendors to specify if the Device Plugins
+can be upgraded without impacting any running containers.
+
+As mentioned earlier, the safest way is to drain the node before upgrading
+the Device Plugins.
+
+# Installation
+
+The installation process should be straightforward to the user, transparent
+and similar to other regular Kubernetes actions.
+The device plugin should also run in containers so that Kubernetes can
+deploy them and restart the plugins when they fail.
+However, we should not prevent the user from deploying a bare metal device
+plugin.
+
+Deploying the device plugins through DemonSets makes sense as the cluster
+admin would be able to specify which machines it wants the device plugins to
+run on, the process is similar to any Kubernetes action and does not require
+to change any parts of Kubernetes.
+
+Additionally, for integrated solutions such as `kubeadm` we can add support
+to auto-deploy community vetted Device Plugins.
+Thus not fragmenting once more the Kubernetes ecosystem.
+
+For users installing Kubernetes without using an integrated solution such
+as `kubeadm` they would use the examples that we would provide at:
+`https://github.com/Kubernetes/Kubernetes/tree/master/examples/device-plugin.yaml`
+
+YAML example:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+spec:
+    template:
+        metadata:
+            labels:
+                - name: device-plugin
+        spec:
+            containers:
+                name: device-plugin-ctr
+                image: NVIDIA/device-plugin:1.0
+                volumeMounts:
+                  - mountPath: /device-plugin
+                  - name: device-plugin
+           volumes:
+             - name: device-plugin
+               hostPath:
+                   path: /var/run/Kubernetes
+```
+
+# Versioning
+
+In today's Kubernetes codebase, there is only one part (CRI) based on
+a protobuf model.
+
+The model used by CRI as of now involves the client (kubelet) checking
+if the server (runtime) version is compatible and then continuing to
+communicate with the server.
+Currently for CRI, compatible means matching the exact version.
+This means that every time the CRI spec changes the CRI clients needs to
+be updated.
+
```

Comment:
I think in general, protobuf is good at handling field additions, so we may not need to bump API version when we add new fields. I would suggest we remove the references on CRI because most the discussions here is general protobuf API principles. We may just say that currently we require exact version match between Kubelet and Device Plugin, and API version is expected to be increased only upon incompatible API changes.